cpp/src/io/parquet/bloom_filter_reader.cu

```diff
@@ -0,0 +1,682 @@
+
+  // Total number of row groups to process.
+  auto const num_row_groups = std::accumulate(
+    input_row_group_indices.begin(),
+    input_row_group_indices.end(),
+    size_t{0},
+    [](size_t sum, auto const& per_file_row_groups) { return sum + per_file_row_groups.size(); });
+
+  // Collect equality literals for each input table column
+  auto const equality_literals =
+    equality_literals_collector{filter.get(), num_input_columns}.get_equality_literals();
+
+  // Collect schema indices of columns with equality predicate(s)
+  std::vector<cudf::size_type> equality_col_schemas;
+  std::for_each(thrust::make_counting_iterator<size_t>(0),
+                thrust::make_counting_iterator(output_column_schemas.size()),
+                [&](auto col_idx) {
+                  // Only for columns that have a non-empty list of literals associated with it
+                  if (equality_literals[col_idx].size()) {
+                    equality_col_schemas.emplace_back(output_column_schemas[col_idx]);
+                  }
+                });
+
+  // Return early if no column with equality predicate(s)
+  if (equality_col_schemas.empty()) { return std::nullopt; }
+
+  // Read a vector of bloom filter bitset device buffers for all column with equality
+  // predicate(s) across all row groups
+  auto const bloom_filter_data = read_bloom_filters(
+    sources, input_row_group_indices, equality_col_schemas, num_row_groups, stream);
+
+  // No bloom filter buffers, return the original row group indices
+  if (bloom_filter_data.empty()) { return std::nullopt; }
+
+  // Get parquet types for the predicate columns
+  auto const parquet_types = get_parquet_types(input_row_group_indices, equality_col_schemas);
+
+  // Copy bloom filter bitset buffer pointers and sizes to device
+  std::vector<void*> h_buffer_ptrs(bloom_filter_data.size());
+  std::vector<size_t> h_buffer_sizes(bloom_filter_data.size());
+  std::for_each(thrust::make_counting_iterator<size_t>(0),
+                thrust::make_counting_iterator<size_t>(bloom_filter_data.size()),
+                [&](auto i) {
+                  auto const& buffer = bloom_filter_data[i];
+                  // Bitset ptr must be non-const to be used in cuco::bloom_filter.
+                  h_buffer_ptrs[i]  = const_cast<void*>(buffer.data());
+                  h_buffer_sizes[i] = buffer.size();
+                });
+  auto buffer_ptrs  = cudf::detail::make_device_uvector_async(h_buffer_ptrs, stream, mr);
+  auto buffer_sizes = cudf::detail::make_device_uvector_async(h_buffer_sizes, stream, mr);
+
+  // Create a bloom filter query table caster.
+  bloom_filter_caster bloom_filter_col{
+    buffer_ptrs, buffer_sizes, parquet_types, num_row_groups, equality_col_schemas.size()};
+
+  // Converts bloom filter membership for equality predicate columns to a table
+  // containing a column for each `col[i] == literal` predicate to be evaluated.
+  // The table contains #sources * #column_chunks_per_src rows.
```

Comment:
Is there a safe way to handle cases where `#sources * #column_chunks_per_src` exceeds the `size_type` maximum row count? Perhaps by disabling bloom filters? Or is that case too esoteric / unlikely to be reached?

Reply:
This is a good question, Imo, it should be quiiite unlikely for (#sources * #column_chunks_per_src = total number of row groups we have), to be > 2B. Maybe @karthikeyann and/or @vuule can better comment on if we should revisit this logic here and in StatsAST or leave as is?

Reply:
For this to be >2B, either #sources should be very large, or `column_chunks_per_src` should be large (which is unlikely). #sources could be very large. A CUDF_EXPECTS should be added to avoid silent failure (at both places). When #sources * #column_chunks_per_src > 2B, won't we exceed `size_type` limit on output table num_rows as well?